src/S3/PostObjectV4.php

```diff
@@ -0,0 +1,209 @@
+        ];
+
+        // setup basic policy
+        $expiration = is_null($expiration)
+                ? gmdate('Y-m-d\TH:i:s\Z', strtotime('+1 hours'))
+                : TimestampShape::format($expiration, 'iso8601');
+
+        $policy = [
+            'expiration' => $expiration,
+            'conditions' => [],
+        ];
+        foreach ($options as $option) {
+            $policy['conditions'][] = $option;
+        }
+
+        $this->formInputs = $formInputs + ['key' => '${filename}'];
+        // add v4 conditions for policy and formInputs
+        foreach ($this->getV4Conditions() as $key => $value) {
+            $policy['conditions'][] = [$key => $value];
+            $this->formInputs += [$key => $value];
+        }
+        $this->jsonPolicy = json_encode($policy);
+
+        // add signature
+        $this->formInputs += $this->getPolicyAndSignature(
+            $this->client->getCredentials()->wait()
+        );
+    }
+
+    /**
+     * Gets the S3 client.
+     *
+     * @return S3ClientInterface
+     */
+    public function getClient()
+    {
+        return $this->client;
+    }
+
+    /**
+     * Gets the bucket name.
+     *
+     * @return string
+     */
+    public function getBucket()
+    {
+        return $this->bucket;
+    }
+
+    /**
+     * Gets the form attributes as an array.
+     *
+     * @return array
+     */
+    public function getFormAttributes()
+    {
+        return $this->formAttributes;
+    }
+
+    /**
+     * Set a form attribute.
+     *
+     * @param string $attribute Form attribute to set.
+     * @param string $value     Value to set.
+     */
+    public function setFormAttribute($attribute, $value)
+    {
+        $this->formAttributes[$attribute] = $value;
+    }
+
+    /**
+     * Gets the form inputs as an array.
+     *
+     * @return array
+     */
+    public function getFormInputs()
+    {
+        return $this->formInputs;
+    }
+
+    /**
+     * Set a form input.
+     *
+     * @param string $field Field name to set
+     * @param string $value Value to set.
+     */
+    public function setFormInput($field, $value)
+    {
+        $this->formInputs[$field] = $value;
+    }
+
+    /**
+     * Gets the raw JSON policy.
+     *
+     * @return string
+     */
+    public function getJsonPolicy()
+    {
+        return $this->jsonPolicy;
+    }
+
+    private function generateUri()
+    {
+        $uri = new Uri($this->client->getEndpoint());
+
+        if ($uri->getScheme() === 'https'
+            && strpos($this->bucket, '.') !== false
+        ) {
+            // Use path-style URLs
+            $uri = $uri->withPath($this->bucket);
+        } else {
+            // Use virtual-style URLs
+            $uri = $uri->withHost($this->bucket . '.' . $uri->getHost());
+        }
+
+        return (string) $uri;
+    }
+
+    protected function getV4Conditions()
+    {
+        $scope = "{$this->sdt}/{$this->client->getRegion()}/s3/aws4_request";
```

Comment:
You can just call `createScope`; it's part of the `SignatureTrait`.

Maybe it would make sense to resolve the region differently if a user injected a multi-region client. You could call `determineBucketRegion` to find out the appropriate region.